### PR TITLE
New arg type RZ_CMD_ARG_TYPE_FOLDER

### LINKED ISSUE #2727: Implement RZ_CMD_ARG_TYPE_FOLDER
Would be nice to have RZ_CMD_ARG_TYPE_FOLDER (including auto completion) which can be very useful.

### PR DESCRIPTION
**Pull request by Arya-1-HR**
**Your checklist for this pull request**
- [x] I've read the [guidelines for contributing](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md) to this repository.
- [x] I made sure to follow the project's [coding style](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md#code-style).
- [ ] I've documented every `RZ_API` function and struct this PR changes.
- [x] I've added tests that prove my changes are effective (required for changes to `RZ_API`).
- [ ] I've updated the [Rizin book](https://github.com/rizinorg/book) with the relevant information (if needed).
- [ ] I've used AI tools to generate fully or partially these code changes and I'm sure the changes are not copyrighted by somebody else.

**Detailed description**

Currently the commands like **cd** , **dtst** , **dtsf** commands are using RZ_CMD_ARG_TYPE_FILE which is wrong , they have the arguments which need to be the path , currently in autocompletion they list also the files as a choice which is wrong

e.g. By pressing TAB after command **cd** it provides the choices of the files to which is not possible.

This PR adds a new arg type **RZ_CMD_ARG_TYPE_FOLDER** which categorise arg types files different and arg type folder/path differently 
and also in autocompletion it only shows the directories to the commands which need only directory path but not files to be suggested 
...

**Test plan**

For this particular case **cd** command will be the best option ,
type command **cd** and press TAB , this will only list you the folders but not the files for autocompletion 
Screenshot 2025-11-09 at 12 00 57 AM

...

**Closing issues**

closes #2727 

...

### CODE REVIEW

File: librz/core/cautocmpl.c [hunk 1]
```diff
@@ -308,6 +308,72 @@ static void autocmplt_cmd_arg_file(RzLineNSCompletionResult *res, const char *s,
 	free(einput);
 }
 
+static void autocmplt_cmd_arg_folder(RzLineNSCompletionResult *res, const char *s, size_t len) {
+	char *input = rz_str_ndup(s, len);
+	if (!input) {
+		return;
+	}
```
Comment (Rot127) [inline librz/core/cautocmpl.c]: You can remove this case.
`RZ_STR_IS_EMPTY` already checks for NULL.

File: librz/core/cautocmpl.c [hunk 2]
```diff
@@ -308,6 +308,72 @@ static void autocmplt_cmd_arg_file(RzLineNSCompletionResult *res, const char *s,
...
+	if (!l) {
+		free(basedir);
+		free(einput);
+		return;
+	}
+
+	RzListIter *iter;
+	char *filename;
+	rz_list_foreach (l, iter, filename) {
+		if (!strcmp(filename, ".") || !strcmp(filename, "..")) {
```
Comment (Rot127) [inline librz/core/cautocmpl.c]: `RZ_STR_EQ` please.

File: librz/core/cautocmpl.c [hunk 3]
```diff
@@ -308,6 +308,72 @@ static void autocmplt_cmd_arg_file(RzLineNSCompletionResult *res, const char *s,
...
+			continue;
+		}
+
+		// Only autocomplete entries that start with the current input
+		if (!strncmp(filename, basename, strlen(basename))) {
```
Comment (Rot127) [inline librz/core/cautocmpl.c]: use `rz_str_startswith_icase`

File: librz/core/cautocmpl.c [hunk 4]
```diff
@@ -308,6 +308,72 @@ static void autocmplt_cmd_arg_file(RzLineNSCompletionResult *res, const char *s,
...
+			char *tmpfilename = rz_file_path_join(basedir, filename);
+
+			// TODO: Only add if it's a directory
```
Comment (Rot127) [inline librz/core/cautocmpl.c]:
```suggestion
```

File: librz/core/cautocmpl.c [hunk 5]
```diff
@@ -308,6 +308,72 @@ static void autocmplt_cmd_arg_file(RzLineNSCompletionResult *res, const char *s,
 	free(einput);
 }
 
+static void autocmplt_cmd_arg_folder(RzLineNSCompletionResult *res, const char *s, size_t len) {
+	char *input = rz_str_ndup(s, len);
```
Comment (Rot127) [inline librz/core/cautocmpl.c]: Would suggest to give variables like these more descriptive names. Like `input-> dir_from_user` or similar.
And `einput -> expanded_path`.
It makes most code way more readable.

File: librz/core/cautocmpl.c [hunk 6]
```diff
@@ -308,6 +308,72 @@ static void autocmplt_cmd_arg_file(RzLineNSCompletionResult *res, const char *s,
...
+
+	char *einput = rz_path_home_expand(input);
+	free(input);
+
+	char *basedir = rz_file_dirname(einput);
+	const char *basename = rz_file_basename(einput + 1);
+
+#if __WINDOWS__
+	rz_str_replace_ch(basedir, '/', '\\', true);
+#endif
```
Comment (Rot127) [inline librz/core/cautocmpl.c]: ~~I think it would be cool to have the whole thing in its own function. Path expansion is pretty common and this way we don't implement it again and again.~~

~~Can you move it into `rz_path_expand` please?~~
~~Also makes the unit testing more focused.~~

File: test/unit/test_cmd.c [hunk 7]
```diff
@@ -1430,6 +1430,38 @@ bool test_call_multiple_macros(void) {
 	mu_end;
 }
 
+bool test_rz_cmd_arg_type_folder_enum(void) {
+	RzCmdArgType type = RZ_CMD_ARG_TYPE_FOLDER;
+	mu_assert_neq(type, RZ_CMD_ARG_TYPE_FILE, "FOLDER should differ from FILE");
+	mu_end;
+}
+
```
Comment (Rot127) [inline test/unit/test_cmd.c]: Not really needed. You can remove it.

**Review by Rot127 — CHANGES_REQUESTED** [review]
Looks good!

You forgot to add the new value here: https://github.com/rizinorg/rizin/blob/ef21a25cc995e2e0e2276f46fd64586554bae254/librz/core/cmd/cmd_api.c#L1409

File: librz/core/cautocmpl.c (hunk 6 above)
Comment (Rot127) [reply librz/core/cautocmpl.c]: Actually, just discovered we have `rz_path_realpath`. Better use this one instead.

**Comment by Rot127** [comment]
Please rebase

File: librz/core/cautocmpl.c (hunk 6 above)
Comment (Arya-1-HR) [reply librz/core/cautocmpl.c]: Was unaware of that , will work on this👍🏻

**Comment by Arya-1-HR** [comment]
Are there any other commands which need this arg types , please refer the files I have changed to check which commands I have identified which need this arg type and already implemented in this PR

**Comment by Rot127** [comment]
> Are there any other commands which need this arg types

From the top of my head I don't know. Please check.

File: test/unit/test_cmd.c [hunk 8]
```diff
@@ -1474,6 +1500,7 @@ int all_tests() {
 	mu_run_test(test_remove_macros);
 	mu_run_test(test_call_macros);
 	mu_run_test(test_call_multiple_macros);
+	mu_run_test(test_cmd_descriptor_folder_arg);
```
Comment (Rot127) [inline test/unit/test_cmd.c:1503]: We still miss a test for the auto-completion. The working directory of those tests should be in `<rizin_repo_root>/test/` (I think, you need to try). So when you do auto-completion for such a command you would expect to get directories in `test/`.

**Review by Rot127 — CHANGES_REQUESTED** [review]

File: librz/core/cautocmpl.c (hunk 6 above)
Comment (Arya-1-HR) [reply librz/core/cautocmpl.c]: As I saw was **rz_path_realpath** need path to be already present on the disk , so it won't work for autocompletion  
like if you say 

e.g.cd ./Doc [TAB]

the Doc as an argument is sent but **rz_path_realpath** returns null , as the path is not already present on disk

**Review by notxvilka — CHANGES_REQUESTED** [review]
```
FAILED: librz/core/rz_core.lib.p/cautocmpl.c.obj 
"cl" "-Ilibrz\core\rz_core.lib.p" "-I." "-I.." "-Ilibrz" "-I..\librz" "-Ilibrz\include" "-I..\librz\include" "-Ilibrz\core\cmd_descs" "-I..\librz\core\cmd_descs" "-Ilibrz\util\sdb\src" "-I..\librz\util\sdb\src" "-I..\librz\arch\isa" "-I..\librz\arch\isa_gnu" "-Ilibrz\arch" "-I..\librz\arch" "-I..\librz\type\parser" "-I..\librz\bin\format\elf" "-I..\librz\bin\format\dmp" "-I..\librz\bin\format\mdmp" "-I..\librz\bin\format\pe" "-I..\librz\bin\format" "-I..\subprojects\rzgdb\include" "-I..\subprojects\rzgdb\include\gdbclient" "-I..\subprojects\rzgdb\include\gdbserver" "-Isubprojects\rzheap" "-I..\subprojects\rzheap" "-I..\subprojects\rzheap\rz_jemalloc" "-I..\subprojects\rzheap\rz_windows" "-I..\subprojects\rizin-shell-parser\src\tree_sitter" "-I..\subprojects\tree-sitter-0.25.3\lib\src" "-I..\subprojects\tree-sitter-0.25.3\lib\include" "/MT" "/nologo" "/showIncludes" "/utf-8" "/W2" "/O2" "/Gw" "-DPSAPI_VERSION=1" "-DRZ_PLUGIN_INCORE=1" "-DSUPPORTS_PCRE2_JIT" "-DZYDIS_STATIC_BUILD" "/Fdlibrz\core\rz_core.lib.p\cautocmpl.c.pdb" /Folibrz/core/rz_core.lib.p/cautocmpl.c.obj "/c" ../librz/core/cautocmpl.c
../librz/core/cautocmpl.c(320): error C2065: 'input': undeclared identifier
../librz/core/cautocmpl.c(320): warning C4047: 'function': 'const char *' differs in levels of indirection from 'int'
../librz/core/cautocmpl.c(320): warning C4024: 'strchr': different types for formal and actual parameter 1
[1999/2408] Compiling C object librz/core/rz_core.lib.p/carg.c.obj
[2000/2408] Compiling C object librz/core/rz_core.lib.p/meson-generated_.._cmd_descs_cmd_descs.c.obj
```

File: test/unit/test_cmd.c (hunk 8 above)
Comment (Arya-1-HR) [reply test/unit/test_cmd.c]: So if I make one function which simulates the autocompletion inside test/ another function should actually verify that autocompletion right?

File: librz/core/cautocmpl.c [hunk 9]
```diff
@@ -308,6 +308,68 @@ static void autocmplt_cmd_arg_file(RzLineNSCompletionResult *res, const char *s,
...
 
+static void autocmplt_cmd_arg_folder(RzLineNSCompletionResult *res, const char *s, size_t len) {
+	char *dir_from_user = rz_str_ndup(s, len);
+
+	if (RZ_STR_ISEMPTY(dir_from_user)) {
+		free(dir_from_user);
+		dir_from_user = rz_str_dup(".");
+	} else if (!rz_file_is_abspath(dir_from_user) && !rz_str_startswith(dir_from_user, ".")) {
+		const char *fmt = ".%s%s";
+#if __WINDOWS__
```
Comment (notxvilka) [inline librz/core/cautocmpl.c]: Some of the code of this function is exactly the same as for `autocmplt_cmd_arg_file`. Rather than duplicating the code, extract parts into small functions and reuse in both places.

**Review by notxvilka — CHANGES_REQUESTED** [review]

File: librz/core/cautocmpl.c (hunk 9 above)
Comment (Arya-1-HR) [reply librz/core/cautocmpl.c]: After adding all that functionality to a function Shall I add it to **autocmplt_cmd_arg_file** too?

Comment (notxvilka) [reply librz/core/cautocmpl.c]: Yes, obviously.

**Comment by notxvilka** [comment]
```
"cl" "-Ilibrz\util\rz_util.lib.p" "-I." "-I.." "-Ilibrz" "-I..\librz" "-Ilibrz\include" "-I..\librz\include" "-Ilibrz\util\sdb\src" "-I..\librz\util\sdb\src" "-Isubprojects\pcre2-10.46" "-I..\subprojects\pcre2-10.46" "-I..\subprojects\pcre2-10.46\src" "-I..\subprojects\softfloat\include" "-Isubprojects\zlib-1.3.1" "-I..\subprojects\zlib-1.3.1" "-I..\subprojects\xz-5.8.1\src\liblzma\api" "/MT" "/nologo" "/showIncludes" "/utf-8" "/W2" "/O2" "/Gw" "-DPSAPI_VERSION=1" "-DRZ_PLUGIN_INCORE=1" "-DSUPPORTS_PCRE2_JIT" "-DZYDIS_STATIC_BUILD" "-DLZMA_API_STATIC" "/Fdlibrz\util\rz_util.lib.p\path.c.pdb" /Folibrz/util/rz_util.lib.p/path.c.obj "/c" ../librz/util/path.c
../librz/util/path.c(336): error C2065: 'dir_from_user': undeclared identifier
../librz/util/path.c(336): warning C4047: 'function': 'const char *' differs in levels of indirection from 'int'
../librz/util/path.c(336): warning C4024: 'strchr': different types for formal and actual parameter 1
[466/2408] Compiling C object librz/util/rz_util.lib.p/pj.c.obj
[467/2408] Compiling C object librz/util/rz_util.lib.p/pkcs7.c.obj
```

**Review by notxvilka — CHANGES_REQUESTED** [review]
It's not the only part of code that is shared between two functions. Please reduce duplication even more.

**Comment by Arya-1-HR** [comment]
> It's not the only part of code that is shared between two functions. Please reduce duplication even more.

The reason because I did this it can be reused again in somewhere else , like path normalisation may be needed in somewhere else too if I extend it can't be used elsewhere and will be constrained only to `rz_cmd_arg_file()` and `rz_cmd_arg_folder()` only

**Comment by notxvilka** [comment]
> > It's not the only part of code that is shared between two functions. Please reduce duplication even more.
> 
> The reason because I did this it can be reused again in somewhere else , like path normalisation may be needed in somewhere else too if I extend it can't be used elsewhere and will be constrained only to `rz_cmd_arg_file()` and `rz_cmd_arg_folder()` only

Extract parts that are common between two functions into small static functions rather than duplicate the code.

File: librz/util/path.c [hunk 10]
```diff
@@ -315,6 +315,41 @@ RZ_API RZ_OWN RzPath *rz_path_new(void) {
...
 }
 
+/**
+ * \brief This function converts arbitrary user input into a safe, normalized path:
+ *  - Empty string ("") is converted to the current directory (".")
+ *  - Relative paths are prefixed with "./"
+ *  - Absolute paths are returned unchanged
+ *  - On Windows, drive-letter paths (e.g. "C:\\") are preserved
+ *
+ * \param usr_input The raw path string provided by the user. May be NULL or empty.
```
Comment (Rot127) [inline librz/util/path.c:325]: `len` is missing.

File: librz/util/path.c [hunk 11]
```diff
@@ -315,6 +315,41 @@ RZ_API RZ_OWN RzPath *rz_path_new(void) {
...
+RZ_API RZ_OWN char *rz_path_normalize_nd_expand(char *usr_input, size_t len) {
+	char *input = rz_str_dup(usr_input);
+	if (RZ_STR_ISEMPTY(input)) {
+		free(input);
+		input = rz_str_dup(".");
+	} else if (!rz_file_is_abspath(input) && !rz_str_startswith(input, ".")) {
+		const char *fmt = ".%s%s";
+#if __WINDOWS__
+		if (strchr(input, ':')) {
+			fmt = "%.0s%s";
```
Comment (Rot127) [inline librz/util/path.c:338]: Why is this formatting needed?
The `.0s` simply seems to not print the `RZ_SYS_DIR`?

File: librz/util/path.c [hunk 12]
```diff
@@ -315,6 +315,41 @@ RZ_API RZ_OWN RzPath *rz_path_new(void) {
...
+ * \brief This function converts arbitrary user input into a safe, normalized path:
+ *  - Empty string ("") is converted to the current directory (".")
+ *  - Relative paths are prefixed with "./"
+ *  - Absolute paths are returned unchanged
+ *  - On Windows, drive-letter paths (e.g. "C:\\") are preserved
+ *
+ * \param usr_input The raw path string provided by the user. May be NULL or empty.
+ * \return A new normalized path string.
+ */
+RZ_API RZ_OWN char *rz_path_normalize_nd_expand(char *usr_input, size_t len) {
```
Comment (Rot127) [inline librz/util/path.c]: Please add unit tests for this function.
Espeicially test for edge cases for paths like
`/home/user/../../etc`, `/home/user/./doc/././`, `.hidden_dir` also for Windows please.

Comment (Rot127) [reply librz/util/path.c]: What does the `_nd_` in the name mean btw?

**Review by Rot127 — CHANGES_REQUESTED** [review]

File: librz/util/path.c (hunk 12 above)
Comment (Arya-1-HR) [reply librz/util/path.c]: > What does the `_nd_` in the name mean btw?

its "and"

File: librz/util/path.c (hunk 10 above)
Comment (Arya-1-HR) [reply librz/util/path.c:325]: Was trying something else , sorry I pushed it without rechecking

File: librz/util/path.c (hunk 12 above)
Comment (Rot127) [reply librz/util/path.c]: Can be removed then.

File: librz/util/path.c (hunk 11 above)
Comment (Arya-1-HR) [reply librz/util/path.c:338]: RZ_SYS_DIR declared with `/` , if that remains then the final path returned will be like

e.g : `/C:\Windows\System32\notepad.exe`

but needed was `C:\Windows\System32\notepad.exe`

Comment (Arya-1-HR) [reply librz/util/path.c:338]: I will add test and reverify again with edge cases , if any problem it will be modified accordingly

**Comment by Rot127** [comment]
```
est_file_slurp OK
test_leading_zeros OK
test_path_prefix OK
test_path_normalize_expand ERR
[XX] Fail at line 88: relative path should be prefixed with './': expected ./foo/bar, got .\foo/bar.
------------------------------------------------------------------------------
112/134 rizin:unit / vector                                OK               0.03s
```

But the result doesn't look right either. On windows it should be all backslashes or not?

**Comment by Arya-1-HR** [comment]
> ```
> est_file_slurp OK
> test_leading_zeros OK
> test_path_prefix OK
> test_path_normalize_expand ERR
> [XX] Fail at line 88: relative path should be prefixed with './': expected ./foo/bar, got .\foo/bar.
> ------------------------------------------------------------------------------
> 112/134 rizin:unit / vector                                OK               0.03s
> ```
> 
> But the result doesn't look right either. On windows it should be all backslashes or not?

yes it should be backslashes , that will be carried out by the autocompletion functions for arg types of folder and file there the `/` will be changed to `\` , @Rot127 you can check this in the `cautocmplt.c` file I have changed

I should look over the case of this block again (sorry for the inconvenience as I don't use windows for test)

**Comment by Rot127** [comment]
> that will be carried out by the autocompletion functions for arg types of folder and file there the `/` will be changed to `\`

This has to be done by `rz_path_normalize_expand`. It is a API function and is therefore used without the auto-complete. Also the function says it normalizes a path. And a path for Windows which uses `/` instead if `\` is not normalized.

File: librz/core/cautocmpl.c [hunk 13]
```diff
@@ -305,7 +289,55 @@ static void autocmplt_cmd_arg_file(RzLineNSCompletionResult *res, const char *s,
 	}
 	rz_list_free(l);
 	free(basedir);
-	free(einput);
+	free(expanded_path);
+}
+
+static void autocmplt_cmd_arg_folder(RzLineNSCompletionResult *res, const char *s, size_t len) {
```
Comment (notxvilka) [inline librz/core/cautocmpl.c]: You still didn't address my suggestion to deduplicate the code

**Review by notxvilka — CHANGES_REQUESTED** [review]

File: librz/core/cautocmpl.c (hunk 13 above)
Comment (Arya-1-HR) [reply librz/core/cautocmpl.c]: Shall I make another function for the rest duplication , cause it will be easier and can combine both `autocmplt_cmd_arg_folder` and `autocmplt_cmd_arg_file` together

File: test/unit/test_util.c [hunk 14]
```diff
@@ -75,10 +75,60 @@ bool test_path_prefix(void) {
...
+	char *out;
+
+	out = rz_path_normalize_expand("", strlen(""));
+	mu_assert_notnull(out, "result should not be NULL");
+	mu_assert_streq(out, ".", "empty input should normalize to '.'");
+	free(out);
+
+	out = rz_path_normalize_expand("foo/bar", strlen("foo/bar"));
+	mu_assert_notnull(out, "result should not be NULL");
+#ifdef __WINDOWS__
```
Comment (notxvilka) [inline test/unit/test_util.c:88]: I am not sure this approach for testing is a proper one. @wargio @Rot127 @kazarmy @b1llow  WDYT?

**Review by notxvilka — CHANGES_REQUESTED** [review]

File: test/unit/test_util.c (hunk 14 above)
Comment (Arya-1-HR) [reply test/unit/test_util.c:88]: Can I know what may go wrong here in you POV @notxvilka

**Review by Rot127 — APPROVED** [review]

File: test/unit/test_util.c (hunk 14 above)
Comment (Rot127) [reply test/unit/test_util.c:88]: This is fine from my point of view. Paths are different on OS, and we don't have unit tests which run exclusively on Windows? (at least I can't find one).

**Review by notxvilka — APPROVED** [review]
Fine then
